packages/@tailwindcss-upgrade/src/template/candidates.ts

```diff
@@ -213,3 +213,51 @@ function simplifyArbitraryVariant(input: string) {
 
   return input
 }
+
+function recursivelyEscapeUnderscores(ast: ValueParser.ValueAstNode[]) {
+  for (let node of ast) {
+    switch (node.kind) {
+      case 'function': {
+        if (node.value === 'url' || node.value.endsWith('_url')) {
+          // Don't decode underscores in url() but do decode the function name
+          node.value = escapeUnderscore(node.value)
+          break
+        }
+
+        if (node.value === 'var' || node.value.endsWith('_var')) {
+          // Don't decode underscores in the first argument of var() but do
+          // decode the function name
+          node.value = escapeUnderscore(node.value)
+          for (let i = 0; i < node.nodes.length; i++) {
+            if (i == 0 && node.nodes[i].kind === 'word') {
+              continue
+            }
+            recursivelyEscapeUnderscores([node.nodes[i]])
+          }
+          break
+        }
+
+        node.value = escapeUnderscore(node.value)
+        recursivelyEscapeUnderscores(node.nodes)
+        break
+      }
+      case 'separator':
+      case 'word': {
+        node.value = escapeUnderscore(node.value)
+        break
+      }
+      default:
+        never()
+    }
+  }
+}
+
+function never(): never {
+  throw new Error('This should never happen')
+}
```

Comment:
The idea here is to use the `never` type so that TypeScript can actually notify us when we add a test (creating an exhaustive switch)

<img width="1011" alt="Screenshot 2024-10-24 at 17 55 37" src="https://github.com/user-attachments/assets/234f6b15-9e6f-48f4-8e1e-1df9dd4ffebc">

But... I did screw this up, we actually need to assert that `node.value` has the type `never` not just the return function here. 😶‍🌫️ 

This way that symbol is never read, it won't even compile if a new type to the ValueParser is added!

Reply:
Ah I tried with `as` and it wasn't working, didn't test satisfies! Yeah might be cleaner.

On the other thing: I think it's fine if it actually builds but at least CI should yell at us if we have a type issue somewhere.